reazon.el

```diff
@@ -164,25 +166,30 @@ STREAM-2, else append them as usual."
           stream-2
           (funcall rest)))))))
 
-(defun reazon--pull (stream)
-  "Force STREAM until it isn't a suspension, then return it."
+(defun reazon--pull (stream &optional stop-time)
+  "Force STREAM until it isn't a suspension or STOP-TIME is reached, then return it."
   (let ((result stream))
     (while (functionp result)
+      (when (and stop-time (< stop-time (float-time)))
+        (signal 'reazon-timeout-error nil))
       (setq result (funcall result)))
     result))
 
-(defun reazon--take (n stream)
-  "Pull N values from STREAM if N is positive else pull it without stopping."
+(defun reazon--take (n stream &optional stop-time)
+  "Pull N values from STREAM if N is positive else pull it without stopping.
+If STOP-TIME is reached, the query will be ended and the values
+collected so far will be returned."
   (declare (indent 1))
   (if (null stream)
       nil
     (let ((count (if n (1- n) -1))
-          (result (list (car stream)))
-          (rest (reazon--pull (cdr stream))))
-      (while (and rest (not (zerop count)))
-        (setq count (1- count))
-        (setq result (cons (car rest) result))
-        (setq rest (reazon--pull (cdr rest))))
+          (result (list (car stream))))
+      (ignore-error 'reazon-timeout-error
+        (let ((rest (reazon--pull (cdr stream) stop-time)))
+          (while (and rest (not (zerop count)))
+            (setq count (1- count))
+            (setq result (cons (car rest) result))
+            (setq rest (reazon--pull (cdr rest) stop-time)))))
```

Comment:
```suggestion
    (let ((count (if n (1- n) -1))
          (result (list (car stream)))
          (rest (ignore-error 'reazon-timeout-error
                  (reazon--pull (cdr stream) stop-time))))
      (while (and rest (not (zerop count)))
        (setq count (1- count))
        (setq result (cons (car rest) result))
        (setq rest (ignore-error 'reazon-timeout-error
                     (reazon--pull (cdr rest) stop-time))))
      (nreverse result))))
```

Reply:
This is written as a `while` loop, but it's really a `do-while` loop. Does Elisp have that? Otherwise there's this ugly duplication.

Reply:
i don't think so unfortunately, but the new exception-less impl. doesn't have duplication anymore